### PR TITLE
Symmetric escape/unescape cookie keys except when it would be a security issue.

### PR DESCRIPTION
**Pull request by ioquatix**
Possible fix for <https://github.com/rack/rack/issues/1865>.

Other possible options:

- Only escape/unescape a subset of characters that cause problems, e.g. "( ) < > @ , ; : \ " / [ ] ? = { }" (i.e. nothing that would cause ambiguity w.r.t. `__Host-` or `__Secure-`).
- Don't escape/unescape keys at all, and instead throw an error when they contain invalid characters.

I personally vote for the latter, but recognise that this PR is more compatible. I think escaping cookie keys is likely to cause confusion as in JavaScript (`document.cookies`) they won't show up in an expected way.

Therefore, I suggest we adopt this PR as a first step, and I'll create a follow up PR which removes escape/unescape entirely - which maybe we can release in a later version of Rack.

### CODE REVIEW

File: lib/rack/utils.rb [hunk 1]
```diff
@@ -211,6 +211,24 @@ def select_best_encoding(available_encodings, accept_encoding)
       (encoding_candidates & available_encodings)[0]
     end
 
+    # A <cookie-name> can be any US-ASCII characters, except control characters, spaces, or tabs. It also must not contain a separator character like the following: ( ) < > @ , ; : \ " / [ ] ? = { }.
+    # Cookies names starting with __Secure- (dash is part of the prefix) must be set with the secure flag from a secure page (HTTPS).
+    # Cookies with names starting with __Host- must be set with the secure flag, must be from a secure page (HTTPS), must not have a domain specified (and therefore aren't sent to subdomains) and the path must be /
```
Comment (jeremyevans) [inline lib/rack/utils.rb:216]: It's odd to have this documentation, but not actually check that these things are true.

**Review by jeremyevans — COMMENTED** [review]
I'll leave the security review of this to @tenderlove.  It does seem odd to me to only have special handling for `__Secure`/`__Host` cookies, though.

**Comment by ioquatix** [comment]
> It does seem odd to me to only have special handling for __Secure/__Host cookies, though.

Those are the cookies that require special handling due to our unescaping logic.

File: lib/rack/utils.rb (hunk 1 above)
Comment (ioquatix) [reply lib/rack/utils.rb:216]: We do a little more validation in the follow up PR.

**Comment by ioquatix** [comment]
Closing in favour of <https://github.com/rack/rack/pull/2191>.

**Pull request closed without merging**